ansi/sixel/palette.go

```diff
@@ -0,0 +1,352 @@
+				case quantizationRed:
+					return cmp.Compare(left.Red, right.Red)
+				case quantizationGreen:
+					return cmp.Compare(left.Green, right.Green)
+				case quantizationBlue:
+					return cmp.Compare(left.Blue, right.Blue)
+				default:
+					return cmp.Compare(left.Alpha, right.Alpha)
+				}
+			})
+
+		// We need to split up the colors in this cube so that the pixels are evenly split between the two,
+		// or at least as close as we can reasonably get. What we do is count up the pixels as we go through
+		// and place the cut point where around half of the pixels are on the left side
+		countSoFar := pixelCounts[uniqueColors[cubeToSplit.startIndex]]
+		targetCount := cubeToSplit.pixelCount / 2
+		leftLength := 1
+
+		for i := cubeToSplit.startIndex + 1; i < cubeToSplit.startIndex+cubeToSplit.length; i++ {
+			c := uniqueColors[i]
+			weight := pixelCounts[c]
+			if countSoFar+weight > targetCount {
+				break
+			}
+			leftLength++
+			countSoFar += weight
+		}
+
+		rightLength := cubeToSplit.length - leftLength
+		rightIndex := cubeToSplit.startIndex + leftLength
+		heap.Push(&cubeHeap, p.createCube(uniqueColors, pixelCounts, cubeToSplit.startIndex, leftLength))
+		heap.Push(&cubeHeap, p.createCube(uniqueColors, pixelCounts, rightIndex, rightLength))
+	}
+
+	// Once we've got max cubes in the heap, pull them all out and load them into the palette
+	for cubeHeap.Len() > 0 {
+		bucketToLoad := heap.Pop(&cubeHeap).(quantizationCube)
+		p.loadColor(uniqueColors, pixelCounts, bucketToLoad.startIndex, bucketToLoad.length)
+	}
+}
+
+// ColorIndex accepts a raw image color (NOT a palette color) and provides the palette index of that color
+func (p *sixelPalette) ColorIndex(c sixelColor) int {
+	return p.paletteIndexes[c]
+}
+
+// loadColor accepts a range of colors representing a single median cut cube. It calculates the
+// average color in the cube and adds it to the palette.
+func (p *sixelPalette) loadColor(uniqueColors []sixelColor, pixelCounts map[sixelColor]uint64, startIndex, cubeLen int) {
+	totalRed, totalGreen, totalBlue, totalAlpha := uint64(0), uint64(0), uint64(0), uint64(0)
+	totalCount := uint64(0)
+	for i := startIndex; i < startIndex+cubeLen; i++ {
+		count := pixelCounts[uniqueColors[i]]
+		totalRed += uint64(uniqueColors[i].Red) * count
+		totalGreen += uint64(uniqueColors[i].Green) * count
+		totalBlue += uint64(uniqueColors[i].Blue) * count
+		totalAlpha += uint64(uniqueColors[i].Alpha) * count
+		totalCount += count
+	}
+
+	averageColor := sixelColor{
+		Red:   uint32(totalRed / totalCount),
+		Green: uint32(totalGreen / totalCount),
+		Blue:  uint32(totalBlue / totalCount),
+		Alpha: uint32(totalAlpha / totalCount),
+	}
+
+	p.PaletteColors = append(p.PaletteColors, averageColor)
+}
+
+// sixelColor is a flat struct that contains a single color: all channels are 0-100
+// instead of anything sensible
+type sixelColor struct {
+	Red   uint32
+	Green uint32
+	Blue  uint32
+	Alpha uint32
+}

```

Comment:
I wonder if it makes more sense if we have `type Color struct { Pc, Pu uint8; Px, Py Pz float64 }` to reflect the color type defined in the Sixel specs which can be a paletted color, RGB (channels are 0-100), or HLS color (Px would be 0-360, Py and Pz are 0-100)

Then we can have helper functions the same as the other types, `func (Color) Encode(io.Writer) error`, `func (Color) String() string`, and `func ParseColor([]byte) (Color, error)`